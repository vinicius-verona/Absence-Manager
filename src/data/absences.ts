import _absences from "./absences.json";

export type Absence = {
    admitterId: number | null;
    admitterNote: string;
    confirmedAt: Date | null;
    createdAt: string;
    crewId: number;
    endDate: string;
    id: number;
    memberNote: string;
    rejectedAt: Date | null;
    startDate: string;
    type: string;
    userId: number;
    status: string;
};

/**
 * Each absence from `absences.json` is inserted into a map.
 *
 * @returns Map<number, Absence> with absence's `id` as key.
 */
export function getAbsences() {
    const absences = new Map<number, Absence>();

    _absences.payload.forEach((_absence) => {
        let _content = _absence as Absence;

        if (!_absence.confirmedAt && !_absence.rejectedAt) {
            _content.status = "Requested";
        } else if (_absence.confirmedAt) {
            _content.status = "Confirmed";
        } else {
            _content.status = "Rejected";
        }

        absences.set(_content.id, _content);
    });

    return absences;
}

export const ABSENCES = getAbsences();
