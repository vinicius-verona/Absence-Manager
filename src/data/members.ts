import _members from "./members.json";

type Member = {
    crewId: number;
    id: number;
    image: string;
    name: string;
    userId: number;
};

/**
 * Each member from `members.json` is inserted into a map.
 *
 * @returns Map<number, Member> with member's `userId` as key.
 */
export function getMembers() {
    const members = new Map<number, Member>();

    _members.payload.forEach((_member) => {
        members.set(_member.userId, _member);
    });

    return members;
}
