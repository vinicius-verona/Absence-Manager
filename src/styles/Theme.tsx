// eslint-disable-next-line import/no-anonymous-default-export
export default {
    palettes: {
        primary: "rgba(255, 148, 25, 1)",
        lightPrimary: "rgba(255, 148, 25, 0.5)",
        light: "#DFDFDF",
        darkerLight: "#B5B5B5",
        bg_dark: "rgba(0, 0, 0, 0.5)",
    },
};
