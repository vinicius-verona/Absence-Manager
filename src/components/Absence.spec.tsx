import { render } from "@testing-library/react";
import { ThemeProvider } from "styled-components";
import Theme from "../styles/Theme";
import AbsenceInfo from "./Absence";
import { MEMBERS } from "../data/members";

describe("Render absence modal on page", () => {
  const absence = {
    admitterId: null,
    admitterNote: "",
    confirmedAt: "2020-12-12T18:03:55.000+01:00",
    createdAt: "2020-12-12T14:17:01.000+01:00",
    crewId: 352,
    endDate: "2021-01-13",
    id: 2351,
    memberNote: "",
    rejectedAt: null,
    startDate: "2021-01-13",
    type: "sickness",
    userId: 2664,
    status: "Confirmed",
  };

  const member = MEMBERS.get(absence.userId);

  if (!member) {
    throw "Member not found";
  }

  it("Should render exit button", () => {
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(getByTestId("exit-button")).toBeInTheDocument();
  });

  it("Should render modal title", () => {
    const { getByText } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(getByText(`Absence ${absence.id}`)).toBeInTheDocument();
  });

  it("Should render all absence's labels and content", () => {
    const { getAllByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    const labels = getAllByTestId("content-label");
    const values = getAllByTestId("content-value");
    expect(labels.length).toBe(5);
    expect(values.length).toBe(5);
  });

  it("Should render absence' status and content", () => {
    const { getByText } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(getByText(absence.status)).toBeInTheDocument();
  });

  it("Should NOT render absence' member note", () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(queryByTestId("member-note")).not.toBeInTheDocument();
  });

  it("Should NOT render absence' admitter note", () => {
    const { queryByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(queryByTestId("admitter-note")).not.toBeInTheDocument();
  });

  it("Should render absence' member note", () => {
    absence.memberNote = "Member note test";
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(getByTestId("member-note")).toBeInTheDocument();
  });

  it("Should render absence' admitter note", () => {
    absence.admitterNote = "Admitter note test";
    const { getByTestId } = render(
      <ThemeProvider theme={Theme}>
        <AbsenceInfo absence={absence} onChange={() => {}} />
      </ThemeProvider>
    );

    expect(getByTestId("admitter-note")).toBeInTheDocument();
  });
});
