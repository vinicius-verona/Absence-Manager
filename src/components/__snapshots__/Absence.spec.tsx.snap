// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render absence modal on page Should match absence modal snapshot 1`] = `
<div
  className="sc-bdvvtL iqTnBm"
>
  <div
    className="sc-gsDKAQ ehuVlV"
  >
    <div
      className="sc-dkPtRN gTDRQT"
    >
      <h2>
        Absence 
        2351
      </h2>
      <span
        className="sc-hKwDye cPklVU"
        data-testid="exit-button"
        onClick={[Function]}
      >
        Ã—
      </span>
    </div>
    <div
      className="sc-eCImPb bThuHP"
    >
      <div
        className="sc-jRQBWg dRDJKa"
      >
        <label>
          <strong
            data-testid="content-label"
          >
            Member: 
          </strong>
        </label>
        <p
          data-testid="content-value"
        >
          Mike
        </p>
      </div>
      <div
        className="sc-jRQBWg dRDJKa"
      >
        <label>
          <strong
            data-testid="content-label"
          >
            User ID: 
          </strong>
        </label>
        <p
          data-testid="content-value"
        >
          2664
        </p>
      </div>
      <div
        className="sc-jRQBWg dRDJKa"
      >
        <label>
          <strong
            data-testid="content-label"
          >
            Type: 
          </strong>
        </label>
        <p
          data-testid="content-value"
        >
          sickness
        </p>
      </div>
      <div
        className="sc-jRQBWg dRDJKa"
      >
        <label>
          <strong
            data-testid="content-label"
          >
            Start Date: 
          </strong>
        </label>
        <p
          data-testid="content-value"
        >
          2021-01-13
        </p>
      </div>
      <div
        className="sc-jRQBWg dRDJKa"
      >
        <label>
          <strong
            data-testid="content-label"
          >
            End Date: 
          </strong>
        </label>
        <p
          data-testid="content-value"
        >
          2021-01-13
        </p>
      </div>
      <div>
        <label>
          <strong>
            Status: 
          </strong>
        </label>
        <label
          className="sc-pVTFL dumrIL"
        >
          Confirmed
        </label>
      </div>
    </div>
    <div
      className="sc-gKclnd iwXWZs centered"
    />
  </div>
</div>
`;
